test/users/npadmana/mpi/test_mpi.chpl

```diff
@@ -0,0 +1,270 @@
+  MPI_Isend(worldRank, 1, MPI_INT, right, toright, MPI_COMM_WORLD, requests[4]);
+
+  MPI_Waitall(4, requests, status);
+
+  writef("Rank %i recieved %i from the left, and %i from the right\n",worldRank, buf[1], buf[2]);
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+/* Compute pi -- test MPI_Reduce */
+proc pi() {
+  const N=10000;
+  var x : [0.. #N] real;
+  var y : [0.. #N] real;
+
+  fillRandom(x); fillRandom(y);
+  var sum = 0.0;
+  forall (x1,y1) in zip(x,y) with (+ reduce sum) {
+    if (x1*x1+y1*y1) <= 1 then sum += 1;
+  }
+  sum /= N;
+
+  var summ : real;
+  MPI_Reduce(sum, summ, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
+
+  if worldRank==0 {
+    summ *= 4/worldSize;
+    writef("I estimate pi to be %r\n",summ);
+  }
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+/* Test scatters. */
+proc test_scatter() {
+  var arr : [{0..3,0..3}]real(32);
+  forall (i,j) in arr.domain {
+    arr[i,j] = (i*4 + j):real(32);
+  }
+  var recbuf : [0..3]real(32);
+
+  // Use MPI Scatter
+  {
+    MPI_Scatter(arr[0,0], 4, MPI_FLOAT, recbuf[0], 4, MPI_FLOAT, 0, MPI_COMM_WORLD);
+    writef("Rank %i :",worldRank); writeln(recbuf);
+    MPI_Barrier(MPI_COMM_WORLD);
+  }
+
+  // Use a contiguous data type
+  {
+    arr += 1:real(32);
+    var rowtype : MPI_Datatype;
+    var stat : MPI_Status;
+
+    MPI_Type_contiguous(4, MPI_FLOAT, rowtype);
+    MPI_Type_commit(rowtype);
+
+    if worldRank==0 {
+      for irank in 0.. #worldSize do MPI_Send(arr[irank,0], 1, rowtype, irank, 1, MPI_COMM_WORLD);
+    }
+    MPI_Recv(recbuf[0], 4, MPI_FLOAT, 0, 1, MPI_COMM_WORLD, stat);
+    writef("Rank %i :",worldRank); writeln(recbuf);
+    MPI_Barrier(MPI_COMM_WORLD);
+    MPI_Type_free(rowtype);
+  }
+
+  // Use a vector data dtype - send columns
+  {
+    var coltype : MPI_Datatype;
+    var stat : MPI_Status;
+
+    MPI_Type_vector(4, 1, 4, MPI_FLOAT, coltype);
+    MPI_Type_commit(coltype);
+
+    if worldRank==0 {
+      for irank in 0.. #worldSize do MPI_Send(arr[0,irank], 1, coltype, irank, 1, MPI_COMM_WORLD);
+    }
+    MPI_Recv(recbuf[0], 4, MPI_FLOAT, 0, 1, MPI_COMM_WORLD, stat);
+    writef("Rank %i :",worldRank); writeln(recbuf);
+    MPI_Barrier(MPI_COMM_WORLD);
+    MPI_Type_free(coltype);
+  }
+
+  // Use an indexed data type
+  {
+    var arr2 = reshape(arr, {0.. #16});
+    var indextype : MPI_Datatype;
+    var stat : MPI_Status;
+
+    var blocklength : [0.. #2]c_int = [2:c_int,2:c_int];
+    var displ : [0.. #2]c_int = [5:c_int,12:c_int];
+
+    MPI_Type_indexed(2, blocklength, displ, MPI_FLOAT, indextype);
+    MPI_Type_commit(indextype);
+
+    if worldRank==0 {
+      for irank in 0.. #worldSize do MPI_Send(arr2[0], 1, indextype, irank, 1, MPI_COMM_WORLD);
+    }
+    MPI_Recv(recbuf[0], 4, MPI_FLOAT, 0, 1, MPI_COMM_WORLD, stat);
+    writef("Rank %i :",worldRank); writeln(recbuf);
+    MPI_Barrier(MPI_COMM_WORLD);
+    MPI_Type_free(indextype);
+  }
+
+}
+
+
+/* Test structure derived type. 
+
+Note that this particular function is deprecated in MPI-1.1
```

Comment:
As far as I can tell, this was due to a refactoring/cleanup of the Fortran bindings. 

I also never use these --- I just send raw bytes over, and assume I know what I'm doing at the other end. I included this example in mostly to demonstrate that even the more involved MPI calls can be used in Chapel....

We may decide to include the MPI-2 updated functions here, if we want.